File: c++/src/leetcode/editor/cn/46-Permutations.cpp
//给定一个 没有重复 数字的序列，返回其所有可能的全排列。 
//
// 示例: 
//
// 输入: [1,2,3]
//输出:
//[
//  [1,2,3],
//  [1,3,2],
//  [2,1,3],
//  [2,3,1],
//  [3,1,2],
//  [3,2,1]
//] 
// Related Topics 回溯算法 
// 👍 985 👎 0

#include<vector>
#include <unordered_set>

using namespace std;

//leetcode submit region begin(Prohibit modification and deletion)
class Solution {
public:
    vector<vector<int>> permute(vector<int> &nums) {
        vector<vector<int>> ans;
        if (nums.size() < 1) { return ans; }
        unordered_set<int> used;
        vector<int> path;
        dfs(nums, 0, used, path, ans);
        return ans;
    }

private:
    void dfs(vector<int> &nums, int depth, unordered_set<int> &used, vector<int> &path, vector<vector<int>> &ans) {
        if (depth == nums.size()) {
            ans.push_back(path);
            return;
        }

        for (int i = 0; i < nums.size(); i++) {
            if (!used.count(nums[i])) {
                used.insert(nums[i]);
                path.push_back(nums[i]);
                dfs(nums, depth + 1, used, path, ans);
                path.pop_back();
                used.erase(nums[i]);
            }
        }
    }
};
//leetcode submit region end(Prohibit modification and deletion)

File: java/src/leetcode/editor/cn/46-Permutations.java
package leetcode.editor.cn;

import java.util.ArrayList;
import java.util.List;

/*
 * @lc app=leetcode.cn id=46 lang=java
 *
 * [46] Permutations
 *
 * https://leetcode-cn.com/problems/permutations/description/
 *
 * algorithms
 * Medium (78.49%)
 * Likes:    2363
 * Dislikes: 0
 * Total Accepted:    783.5K
 * Total Submissions: 993.5K
 * Testcase Example:  '[1,2,3]'
 *
 * Given an array nums of distinct integers, return all the possible
 * permutations. You can return the answer in any order.
 * 
 * 
 * Example 1:
 * Input: nums = [1,2,3]
 * Output: [[1,2,3],[1,3,2],[2,1,3],[2,3,1],[3,1,2],[3,2,1]]
 * Example 2:
 * Input: nums = [0,1]
 * Output: [[0,1],[1,0]]
 * Example 3:
 * Input: nums = [1]
 * Output: [[1]]
 * 
 * 
 * Constraints:
 * 
 * 
 * 1 <= nums.length <= 6
 * -10 <= nums[i] <= 10
 * All the integers of nums are unique.
 * 
 * 
 */

// @lc code=start
class Solution {
    public List<List<Integer>> permute(int[] nums) {
        if (nums.length < 1) {
            return ans;
        }

        backtrack(nums, new ArrayList<>(), new boolean[nums.length]);
        return ans;
    }

    private List<List<Integer>> ans = new ArrayList<>();

    private void backtrack(int[] nums, List<Integer> path, boolean[] used) {
        if (path.size() == nums.length) {
            ans.add(new ArrayList<>(path));
            return;
        }

        for (int i = 0; i < nums.length; i++) {
            if (!used[i]) {
                used[i] = true;
                path.add(nums[i]);
                backtrack(nums, path, used);
                used[i] = false;
                path.remove(path.size() - 1);
            }
        }
    }
}
// @lc code=end
